File: android/lan-gaming/src/main/kotlin/com/locxx/langaming/LanGamingClient.kt
package com.locxx.langaming

import android.os.Handler
import android.os.Looper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import kotlin.coroutines.coroutineContext

/**
 * HTTP client: POST [CLIENT_HELLO] to `/locxx/v1/hello`, then long-poll `/locxx/v1/poll` and POST
 * frames to `/locxx/v1/send` (see docs/protocol.md).
 */
class LanGamingClient(
    private val listener: LanGamingListener
) {
    private val mainHandler = Handler(Looper.getMainLooper())
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var runJob: Job? = null

    var displayName: String = "Player"

    private var baseUrl: String? = null
    private var token: String? = null

    fun connect(hostBaseUrl: String) {
        disconnect()
        val normalized = hostBaseUrl.trimEnd('/')
        baseUrl = normalized
        val name = displayName
        runJob = scope.launch(Dispatchers.IO) {
            try {
                helloAndRun(normalized, name)
            } catch (e: kotlinx.coroutines.CancellationException) {
                throw e
            } catch (e: Exception) {
                onErrorUi(e.message ?: "client")
            }
        }
    }

    private suspend fun helloAndRun(base: String, name: String) {
        val nonce = ByteArray(16).also { java.security.SecureRandom().nextBytes(it) }
        val hello = ProtocolCodec.buildClientHello(WIRE_VERSION, name, nonce)
        val conn = (URL("$base/locxx/v1/hello").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            setRequestProperty("Content-Type", "application/octet-stream")
            doOutput = true
            connectTimeout = 10_000
            readTimeout = 10_000
        }
        conn.outputStream.use { it.write(hello) }
        val code = conn.responseCode
        if (code != 200) {
            onErrorUi("hello http $code")
            conn.disconnect()
            return
        }
        val tok = conn.getHeaderField(LOCXX_HTTP_TOKEN_HEADER) ?: run {
            onErrorUi("missing token")
            conn.disconnect()
            return
        }
        val welcome = conn.inputStream.use { it.readBytes() }
        conn.disconnect()

        val decoded = ProtocolCodec.decodeFrame(welcome).getOrElse {
            onErrorUi("welcome decode")
            return
        }
        if (decoded.messageType != WireMessageType.HOST_WELCOME) {
            onErrorUi("expected welcome")
            return
        }
        val w = ProtocolCodec.parseHostWelcome(decoded.payload).getOrElse {
            onErrorUi("welcome parse")
            return
        }
        token = tok
        onPeerConnectedUi(tok, w.playerId.toInt(), name)

        while (coroutineContext.isActive && token == tok) {
            val pollConn = (URL(
                "$base/locxx/v1/poll?token=${
                    URLEncoder.encode(
                        tok,
                        Charsets.UTF_8.name()
                    )
                }"
            ).openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                connectTimeout = 10_000
                readTimeout = 120_000
            }
            val pollCode = try {
                pollConn.responseCode
            } catch (e: Exception) {
                onErrorUi(e.message ?: "poll")
                pollConn.disconnect()
                break
            }
            if (pollCode == HttpURLConnection.HTTP_NO_CONTENT) {
                pollConn.disconnect()
                continue
            }
            if (pollCode != 200) {
                onErrorUi("poll http $pollCode")
                pollConn.disconnect()
                break
            }
            val body = pollConn.inputStream.use { it.readBytes() }
            pollConn.disconnect()
            if (body.isEmpty()) continue
            val pollResult = ProtocolCodec.decodeFrame(body)
            if (pollResult.isFailure) {
                onErrorUi("poll decode")
                continue
            }
            val pollDecoded = pollResult.getOrThrow()
            when (pollDecoded.messageType) {
                WireMessageType.PING -> sendFrame(ProtocolCodec.encodeFrame(WireMessageType.PONG))
                else -> onMessageReceivedUi(tok, pollDecoded.messageType, pollDecoded.payload)
            }
        }
        token = null
        baseUrl = null
        if (coroutineContext.isActive) {
            onPeerDisconnectedUi(tok)
        }
    }

    fun sendFrame(frame: ByteArray) {
        val b = baseUrl ?: return
        val tok = token ?: return
        scope.launch(Dispatchers.IO) {
            runCatching {
                val conn = (URL(
                    "$b/locxx/v1/send?token=${
                        URLEncoder.encode(
                            tok,
                            Charsets.UTF_8.name()
                        )
                    }"
                ).openConnection() as HttpURLConnection).apply {
                    requestMethod = "POST"
                    setRequestProperty("Content-Type", "application/octet-stream")
                    doOutput = true
                    connectTimeout = 10_000
                    readTimeout = 30_000
                }
                conn.outputStream.use { it.write(frame) }
                val c = conn.responseCode
                conn.disconnect()
                if (c != HttpURLConnection.HTTP_NO_CONTENT) {
                    onErrorUi("send http $c")
                }
            }.onFailure { e ->
                onErrorUi(e.message ?: "send")
            }
        }
    }

    fun disconnect() {
        val t = token
        runJob?.cancel()
        runJob = null
        token = null
        baseUrl = null
        if (t != null) {
            mainHandler.post { listener.onPeerDisconnected(t) }
        }
    }

    private fun onPeerConnectedUi(addr: String, pid: Int, name: String) {
        mainHandler.post { listener.onPeerConnected(addr, pid, name) }
    }

    private fun onPeerDisconnectedUi(addr: String) {
        mainHandler.post { listener.onPeerDisconnected(addr) }
    }

    private fun onMessageReceivedUi(addr: String, type: Byte, payload: ByteArray) {
        mainHandler.post { listener.onMessageReceived(addr, type, payload) }
    }

    private fun onErrorUi(msg: String) {
        mainHandler.post { listener.onError(msg) }
    }
}

File: ios/LanGaming/Sources/LanGaming/LanGamingClient.swift
import Foundation
import Security

public final class LanGamingClient {
    private let listener: LanGamingListener
    private var task: Task<Void, Never>?
    public var displayName: String = "Player"
    private var token: String?
    private var baseUrl: String?

    public init(listener: LanGamingListener) {
        self.listener = listener
    }

    public func connect(hostBaseUrl: String) {
        disconnect()
        let trimmed = hostBaseUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        var base = trimmed.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if !base.lowercased().hasPrefix("http") {
            base = "http://\(base)"
        }
        baseUrl = base
        task = Task { [weak self] in
            guard let self else { return }
            await self.run(base: base)
        }
    }

    public func disconnect() {
        task?.cancel()
        task = nil
        let t = token
        token = nil
        baseUrl = nil
        if let t {
            DispatchQueue.main.async {
                self.listener.onPeerDisconnected(address: t)
            }
        }
    }

    public func sendFrame(_ frame: Data) {
        guard let base = baseUrl, let tok = token else { return }
        let enc = tok.addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed) ?? tok
        guard let u = URL(string: "\(base)/locxx/v1/send?token=\(enc)") else { return }
        Task {
            var req = URLRequest(url: u)
            req.httpMethod = "POST"
            req.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
            req.httpBody = frame
            do {
                let (_, resp) = try await URLSession.shared.data(for: req)
                guard let http = resp as? HTTPURLResponse else { return }
                if http.statusCode != 204 {
                    DispatchQueue.main.async {
                        self.listener.onError("send http \(http.statusCode)")
                    }
                }
            } catch {
                if self.isCancelledLike(error) { return }
                DispatchQueue.main.async {
                    self.listener.onError(error.localizedDescription)
                }
            }
        }
    }

    private func run(base: String) async {
        do {
            var nonce = Data(count: 16)
            _ = nonce.withUnsafeMutableBytes { ptr in
                SecRandomCopyBytes(kSecRandomDefault, 16, ptr.baseAddress!)
            }
            let hello = try ProtocolCodec.buildClientHello(protocolVersion: wireVersion, displayName: displayName, clientNonce: nonce)
            var req = URLRequest(url: URL(string: "\(base)/locxx/v1/hello")!)
            req.httpMethod = "POST"
            req.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
            req.httpBody = hello
            let (welcomeData, response) = try await URLSession.shared.data(for: req)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                DispatchQueue.main.async { self.listener.onError("hello http") }
                return
            }
            let hdr = http.value(forHTTPHeaderField: locxxHttpTokenHeader) ?? http.value(forHTTPHeaderField: "X-LocXX-Token")
            guard let assigned = hdr else {
                DispatchQueue.main.async { self.listener.onError("missing token") }
                return
            }
            let decoded = try ProtocolCodec.decodeFrame(welcomeData)
            guard decoded.messageType == WireMessageType.hostWelcome else {
                DispatchQueue.main.async { self.listener.onError("expected welcome") }
                return
            }
            let w = try ProtocolCodec.parseHostWelcome(decoded.payload)
            let tok = assigned
            token = tok
            DispatchQueue.main.async {
                self.listener.onPeerConnected(address: tok, playerId: Int(w.playerId), displayName: self.displayName)
            }
            try await pollLoop(base: base, tok: tok)
            token = nil
            baseUrl = nil
            if !Task.isCancelled {
                DispatchQueue.main.async {
                    self.listener.onPeerDisconnected(address: tok)
                }
            }
        } catch {
            token = nil
            baseUrl = nil
            if isCancelledLike(error) { return }
            DispatchQueue.main.async {
                self.listener.onError(error.localizedDescription)
            }
        }
    }

    private func isCancelledLike(_ error: Error) -> Bool {
        error is CancellationError || (error as? URLError)?.code == .cancelled
    }

    private func pollLoop(base: String, tok: String) async throws {
        let allowed = CharacterSet.urlQueryAllowed
        while !Task.isCancelled {
            let enc = tok.addingPercentEncoding(withAllowedCharacters: allowed) ?? tok
            guard let u = URL(string: "\(base)/locxx/v1/poll?token=\(enc)") else { break }
            var req = URLRequest(url: u)
            req.httpMethod = "GET"
            req.timeoutInterval = 120
            let (data, response) = try await URLSession.shared.data(for: req)
            guard let http = response as? HTTPURLResponse else { break }
            if http.statusCode == 204 { continue }
            guard http.statusCode == 200 else {
                DispatchQueue.main.async { self.listener.onError("poll http \(http.statusCode)") }
                break
            }
            if data.isEmpty { continue }
            let decoded = try ProtocolCodec.decodeFrame(data)
            switch decoded.messageType {
            case WireMessageType.ping:
                sendFrame(ProtocolCodec.encodeFrame(messageType: WireMessageType.pong))
            default:
                DispatchQueue.main.async {
                    self.listener.onMessageReceived(address: tok, messageType: decoded.messageType, payload: decoded.payload)
                }
            }
        }
    }
}
